substrate/frame/message-queue/src/lib.rs

```diff
@@ -1447,7 +1449,22 @@ impl<T: Config> Pallet<T> {
 		use ProcessMessageError::*;
 		let prev_consumed = meter.consumed();
 
-		match T::MessageProcessor::process_message(message, origin.clone(), meter, &mut id) {
+		let transaction =
+			storage::with_transaction(|| -> TransactionOutcome<Result<_, DispatchError>> {
+				let res =
+					T::MessageProcessor::process_message(message, origin.clone(), meter, &mut id);
+				match &res {
+					Ok(_) => TransactionOutcome::Commit(Ok(res)),
+					Err(_) => TransactionOutcome::Rollback(Ok(res)),
+				}
+			});
+
+		let transaction = match transaction {
+			Ok(result) => result,
+			_ => return MessageExecutionStatus::Unprocessable { permanent: false },
```

Comment:
```suggestion
			_ => return MessageExecutionStatus::Unprocessable { permanent: true },
```

This should also use the `defensive!` macro to log an error since it is very much unexpected.